keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -230,80 +234,177 @@ scenario of wrong-single-stack pods...
+"endpoint reconciler" modes: `master-count` (the old algorithm), and
+`lease` (the current default, and preferred mode).
+
+In `master-count` mode, each apiserver periodically checks the
+`kubernetes` `Endpoints` object, and:
+
+  - If its own IP is not listed there, it adds it.
+
+  - If the total number of IPs listed there is now greater than the
+    expected number of masters, it removes a randomly-selected IP
+    other than its own.
+
+(The changes to `Endpoints` are likewise synced to `EndpointSlice`.)
+
+This means that when an apiserver is replaced, it may take arbitrarily
+long for the `Endpoints` to settle on the correct values (since none
+of the remaining apiservers knows which IP is the one that needs to be
+removed), but once it reaches the correct values, it will stay correct
+until the set of apiservers changes again.
+
+In the `lease` mode, each apiserver periodically writes its IP to a
+non-kubernetes-API object in etcd, along with a timestamp and a lease
+time. Then it reads all of the other IPs there, and updates the
+`Endpoints` and `EndpointSlice` to contain all of the IPs whose leases
+have not expired. So in this mode, a stale IP is guaranteed to be
+removed once its lease expires, and no apiserver will ever remove a
+valid IP from the list.
+
+#### Upgrading from current apiservers to new dual-stack apiservers
+
+As of 1.22, apiservers only reconcile the `Service` object at startup,
+and only ever reconcile a single `EndpointSlice` object for the
+primary IP family. The latter behavior is clearly a bug
+([kubernetes #101070]), and it has been suggested that the former
+behavior is a bug as well. However, if we avoid fixing these bugs
+until we add dual-stack-apiserver support, then it makes upgrades from
+old/current apiservers to future dual-stack apiservers easier:
+
+  - When the first dual-stack apiserver comes up, it would switch the
+    `kubernetes` Service to dual-stack, and then periodically
+    reconcile that.
+
+    Old apiservers would only ever reconcile the Service on startup,
+    so even if an old apiserver restarted during the upgrade process
+    and reverted the Service back to single-stack, it would get
+    re-reverted back to dual-stack by one of the running dual-stack
+    apiservers eventually.
+
+  - When the first dual-stack apiserver comes up, it would add its
+    primary IP to the `kubernetes` Endpoints and EndpointSlice, and
+    create a secondary EndpointSlice with its secondary IP.
+
+    Old apiservers would agree that the new apiserver's primary IP
+    belonged in the Endpoints and primary EndpointSlice, and would
+    ignore the secondary EndpointSlice, so we would get the right
+    result here as well.
+
+  - Eventually all of the apiservers would be running the new code,
+    and would all cooperatively maintain the dual-stack Service and
+    dual-stack EndpointSlices.
+
+The downgrade case will also (mostly) work, as long as the last of the
```

Comment:
So in that case, when the first apiserver converts to single-stack, it would remove its own secondary-IP-family endpoint from the pool, but it would see that the other two apiservers were still advertising dual-stack endpoints, so it would leave the Service as dual-stack. Then the second apiserver would do the same thing. When the last apiserver converts to single-stack, it would see that there were no longer any dual-stack endpoints for the apiserver service, so it would flip the Service to single-stack.

Reply:
I think I like that